File: reverse.py
#!/usr/bin/env nix-shell
#! nix-shell -i python3 -p python3

import math
import sys

def digits(n):
    return math.ceil(math.log10(n))

def reverse(n):
    m = 0
    i = digits(n) - 1
    for i in range(i,-1,-1):  # [i,i-1..0]
        (n,r) = divmod(n, 10)
        m += r*10**i
    return m

if __name__ == "__main__":
    n = int(sys.argv[1])
    output = "{} -> {}".format(n, reverse(n))
    print(output)

File: reverse.hs
#!/usr/bin/env nix-shell
#! nix-shell -i runghc -p myHaskellEnv

module Main where

import System.Environment
import Control.Monad
import Control.Monad.ST.Strict
import Data.STRef

main = do
  (n:_) <- getArgs
  let n' = read n :: Integer
  putStrLn $ show n' ++ " -> " ++ show (reverse' n')

infix 0 |$=|
(|$=|) = modifySTRef
infix 0 |:=|
(|:=|) = writeSTRef

reverse' :: (Show a, Integral a) => a -> a
reverse' n = runST $ do
  n <- newSTRef n
  m <- newSTRef 0
  forM_ [i,i-1..0] $ \i -> do
    n' <- readSTRef n
    let (q,r) = n' `divMod` 10
    m |$=| (+ r*10^i)
    n |:=| q
  readSTRef m

  where i = (digits n)-1

digits :: Integral a => a -> a
digits = ceiling . logBase 10 . fromIntegral
